### PR TITLE
Check public hosts' SSL expiry

### PR DESCRIPTION
**Pull request by gidsg**
Code borrowed from https://github.com/guardian/frontend/blob/master/integrated-tests/src/test/scala/common/SslCertTest.scala 

I have not included development environment certificates as it would involve a load of hassle to import the GNM Cert into the Trusted Certificate Store.

CC @guardian/content-platforms

### CODE REVIEW

File: app/com/gu/contentapi/sanity/SSLExpiryTest.scala [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    handleException {
+      val hosts = Seq (
+        Config.host,
+        Config.hostPublicSecure,
+        Config.previewHost,
+        Config.writeHost,
+        Config.writePreviewHost
+      )
+
+      val secureHosts = for (host <- hosts) yield host.replaceAll("http://", "https://")
```
Comment (Nick-Smith) [inline app/com/gu/contentapi/sanity/SSLExpiryTest.scala]: Use map to do this more succinctly:
val secureHosts = hosts map {_.replaceAll("http://", "https://")}

**Comment by Nick-Smith** [comment]
A few points:

1. You MUST call conn.disconnect even if you throw an exception! Otherwise you will run out of connections.
Either put this line in a finally{ } block or see below for Try.

2. Rather than using try{] catch{} which is sucky, use scala's Try{}. 

3. You are duplicating lots of code around the exceptions. You can handle all these exceptions with one line of code by using:
case e: Throwable => fail(s"Some exception" + e.getClass.getSimpleName)

As an example of how I would write this using Try:

val conn = url.openConnection().asInstanceOf[HttpsURLConnection]
conn.setHostnameVerifier(new HostnameVerifier {
  override def verify(hostnameVerifier: String, sslSession: SSLSession) = true
}
val certsTry = Try { 
  conn.connect()
  conn.getServerCertificates
}
conn.disconnect()
certsTry match {
  case Success(certs) => // Test the certs expiry

  case Failure(e) => fail(s"Exception thrown: ${e.getClass.getSimpleName}")
}

If you need any help just let me know.

**Comment by LATaylor-guardian** [comment]
Nice explanation @Nick-Smith! This is a problem we've had to fix a few times in the indexer where the connection is not released.

**Comment by mchv** [comment]
Please do not catch `NullPointerException` or `Throwable` directly but specifically the exceptions you target. 
```scala
catch {
  case e @ (_: MappingException | _: ParseException | _: SomeOtherException) =>
}
```

**Comment by gidsg** [comment]
Thanks @Nick-Smith I have made the change suggested, however, if the connection fails for whatever reason I call Cancel instead of Fail as I don't want to trigger an alert.

**Comment by Nick-Smith** [comment]
@mchv  Yes the Neophyte's guide has some great tutorials on Scala.

I haven't seen that syntax for matching multiple types before, very handy!

It's probably worth mentioning you shouldn't do your assertions inside the Try{} block - failed assertions are Exceptions so they will be caught by the Try and the test won't fail.

**Comment by gidsg** [comment]
@Nick-Smith does this look OK to merge now?

File: app/com/gu/contentapi/sanity/SSLExpiryTest.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        val certsTry = Try {
+          conn.connect()
+          conn.getServerCertificates
+        }
+        conn.disconnect()
+
+        certsTry match {
+          case Success(certs) =>
+            certs.headOption.map { cert =>
+              val x = cert.asInstanceOf[X509CertImpl]
```
Comment (cb372) [inline app/com/gu/contentapi/sanity/SSLExpiryTest.scala]: Not a big problem (I'd say it's fine to merge as-is), but before calling `asInstanceOf` I would usually have an assertion to check that the thing is actually what I am trying to cast it to.

```
certs.headOption.map { cert =>
  cert shouldBe a[X509CertImpl]
  val x = cert.asInstanceOf[X509CertImpl]
  ...
}
```

File: app/com/gu/contentapi/sanity/SSLExpiryTest.scala [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+        val certsTry = Try {
+          conn.connect()
+          conn.getServerCertificates
+        }
+        conn.disconnect()
+
+        certsTry match {
+          case Success(certs) =>
+            certs.headOption.map { cert =>
```
Comment (cb372) [inline app/com/gu/contentapi/sanity/SSLExpiryTest.scala]: The test will pass (by doing nothing) if the list of certificates is empty. Is that expected?

File: app/com/gu/contentapi/sanity/SSLExpiryTest.scala (hunk 2 above)
Comment (gidsg) [inline app/com/gu/contentapi/sanity/SSLExpiryTest.scala]: Thanks, have implemented this check now.

File: app/com/gu/contentapi/sanity/SSLExpiryTest.scala (hunk 3 above)
Comment (gidsg) [inline app/com/gu/contentapi/sanity/SSLExpiryTest.scala]: Good idea, I check now that there is at least 1 certificate in the array.

**Comment by gidsg** [comment]
OK, @cb372 I have made the suggested changes so will merge now,

**Pull request merged**